### PR TITLE
Fix linker flags for -static-executable

### PR DESCRIPTION
**Pull request by drexin**
Also adds a test to ensure it does not break in the future

Building static executables on Linux is currently broken because the ICU libraries in static-executable-args.lnk are missing the swift postfix.

### CODE REVIEW

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
One thing I'm not sure about is how to handle other corelibs like Foundation and Dispatch. Today we don't link against them by default, so users would have to pass in linker flags to make it work (including all their respective dependencies). Is there any harm in just always passing the right flags for those libraries, in case they are needed? The linker should be smart enough to not include them when they aren't being used.

**Comment by drexin** [comment]
@swift-ci smoke test macos

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by compnerd** [comment]
No, passing them in means that they must be present - that may or may not be true.  Autolinking should actually handle that case though - `import Foundation` should cause `-lFoundation` to be injected and cause that to get linked in.

File: lib/Driver/UnixToolChains.cpp [hunk 1]
```diff
@@ -212,7 +212,9 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
   }
 
   if (getTriple().getOS() == llvm::Triple::Linux &&
-      job.getKind() == LinkKind::Executable) {
+      job.getKind() == LinkKind::Executable &&
+      !context.Args.hasFlag(options::OPT_static_executable,
+                            options::OPT_no_static_executable, false)) {
     Arguments.push_back("-pie");
```
Comment (compnerd) [inline lib/Driver/UnixToolChains.cpp:218]: Not really your fault, but this seems that it is an ELF specific flag, Android and Linux both support this, as do other platforms like BSDs and Haiku IIRC.  I'm not sure that there is any value to the OS check.

File: test/Driver/static-executable.swift [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Create a self contained binary
+// REQUIRES: OS=linux-gnu
+// REQUIRES: static_stdlib
+print("hello world!")
+// RUN: %empty-directory(%t)
+// RUN: %target-swiftc_driver -static-executable -o %t/static-executable %s
+// RUN: %t/static-executable | %FileCheck %s
+// RUN: not ldd %t/static-executable | %FileCheck %s --check-prefix=LDD
```
Comment (compnerd) [inline test/Driver/static-executable.swift:8]: Please do not use `ldd` - that is a security issue.  Please use `llvm-readelf` instead to check the binary type.

**Review by compnerd — CHANGES_REQUESTED** [review]

File: lib/Driver/UnixToolChains.cpp (hunk 1 above)
Comment (drexin) [reply lib/Driver/UnixToolChains.cpp:218]: Ok, I'll fix that.

File: test/Driver/static-executable.swift (hunk 2 above)
Comment (drexin) [reply test/Driver/static-executable.swift:8]: Ah, thanks!

**Comment by drexin** [comment]
> No, passing them in means that they must be present - that may or may not be true. Autolinking should actually handle that case though - `import Foundation` should cause `-lFoundation` to be injected and cause that to get linked in.

Yeah, that's what I thought. Unfortunately autolinking does not seem to work properly for static executables. Looking more closely it seems like it does pass `-lFoundation`, but none of its dependencies, so I'm getting missing symbols for CoreFoundation. Any idea how to fix that?

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test macos

**Comment by drexin** [comment]
@swift-ci smoke windows

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by compnerd** [comment]
Yes, I do know how to fix it, its not pretty though.  Because I was unable to switch the builds to lld here since we dont build it, I lost the motivation to finish the lld support that would allow that.  We would need to revive that work and get that merged to support this.

**Review by compnerd — APPROVED** [review]

**Pull request merged**